jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java

```diff
@@ -286,6 +286,29 @@ public static void putIntLittleEndian(ByteBuffer buffer, int value)
         flipToFlush(buffer, p);
     }
 
+    /**
+     * Slice a buffer given an offset and a length.
+     * @param buffer the buffer to slice
+     * @param offset the offset
+     * @param length the length, -1 meaning use the current limit
+     * @return the sliced buffer
+     */
+    public static ByteBuffer slice(ByteBuffer buffer, int offset, int length)
+    {
+        ByteBuffer slice = buffer.slice();
+        if (offset > 0)
+        {
+            int newPosition = slice.position() + offset;
+            if (newPosition > slice.limit() && length == 0)
+                slice.position(slice.limit());
+            else
+                slice.position(newPosition);
+        }
+        if (length > -1)
```

Comment:
This doesn't check for length < -1, and will not do the right thing for `length==0` when the existing `ByteBuffer` already has a length set.

Reply:
When `length` is 0, the limit is set to the position. How is that incorrect?

Reply:
sorry, I meant when the existing buffer already has a _position_ set.
If offset = 0, position is unchanged.

Reply:
`offset` really is an offset and not an index, i.e.: it's relative to the current position.